### ISSUE TITLE
Support for acyclic directed relationships

### ISSUE THREAD

**Issue by Raznic** [neo4j-contrib/neomodel#252]
When defining a directed relationship on a StructuredNode class (i.e. RelationshipTo, RelationshipFrom) it would be nice to have the option to specify that the relationship should be acyclic. 

As a use case, let's say we have the following class defined:
```python
class Person(StructuredNode):
   ...
   parents = RelationshipTo('Person', 'PARENT')
```
The above StructuredNode class defines a Person with a relationship called `parents`.  In most cases, we would want to prevent a person from becoming their own ancestor. To accomplish this, we would need to ensure that each time a node is connected to a Person through the `parents` relationship, it won't create a cycle.

I've cloned the neomodel repository and taken an initial crack at the implementation (Raznic@d68f1174903879207763c19674b1da3c906a24f7). Before I submitted a pull request I wanted to see if there would be any interest in getting this feature back into the main project, or if there were any other suggestions for the feature.

**Comment by robinedwards** [comment]
Super made a comment on the commit. This is brilliant that you.

**Comment by croccam** [comment]
To overcome the race condition: can't you just check if the start.id and end.id are the same without any cypher query? Since you have self.id and node.id, you would just have to check if they are equal, right? Or am I missing something out?

Regards.

**Comment by Raznic** [comment]
@croccam thanks for the input! That is one way to check when we create a relationship from a node back to itself. That is definitely an edge case and something I need to check for. I think in most use cases the nodes will be different, so the ID check wouldn't work for those instances. We would still need to run the Cypher query to see if there is currently a path from the end node back to the start. If so, adding a relationship from the start node to the end node would introduce a cycle. In order to overcome the race condition, I wrapped the check/relationship creation inside of a database transaction. However, I need to check if there is currently a transaction already taking place before starting a new one. Sorry for the lack of forward progress on this, hoping to make some movement on it this week.

**Comment by croccam** [comment]
My bad, I read it so fast that I misunderstood your original issue. I get it now, thanks for your answer. Transactions seem like a good solution. I would also be really interested on how to overcome the possible performance issues if there were too many long branches to check, maybe not for the parent use case, but for other use cases.

**Comment by Raznic** [comment]
Yeah, I haven't tested against any large databases, so I'm not sure how it performs once you grow beyond a simple development instance. If you're aware of any place I could get my hands on some test data, let me know. I'd definitely be interested to see how it handles under load.

**Comment by aanastasiou** [comment]
@Raznic Has there been any progress with this?

I think that it would be worth adding a `Relationship` parameter like `prevent_loops` but that would cover [cycles of length zero](https://en.wikipedia.org/wiki/Loop_(graph_theory)) **only**. Avoiding _cycles_ would be a much more difficult check to perform at the level of `neomodel`.

The loop check can be performed on asserting that the `start_node` and `end_node` have different node ids.

Can you please let me know if that was the original intention?

**Comment by Raznic** [comment]
Sorry for the delayed reply. So the original intention was to detect cycles of any length. The plan was to check if there was a path from the `end_node` to the `start_node`. If that path existed, then adding the relationship from `start_node` to `end_node` would create a cycle and the code would raise an exception. It was pretty simple to implement, but the part I got hung up on was how to do this in an atomic fashion. You would need to ensure that no new relationships can be added during the transaction of checking for the `end_node` to `start_node` path and creating the relationship. If you do, you may run in to a race condition that ends up creating a cycle. 

Shortly after reaching this conundrum, the project that I was working on using nemodel switched over to a different database, so I never really swung back to complete this feature. If anyone wants to pick up development and finish this feature, they're more than welcome to take a look at the work I've already done on this [branch](https://github.com/Raznic/neomodel/tree/feature_acyclic_relationships). Just a heads up, I haven't merged with the official project in almost 2 years, so it's probably well out-of-date. Still, hopefully it can provide some insight on at least getting started.

**Comment by aanastasiou** [comment]
@Raznic Thank you very much for coming back to us about this. Having worked on the problem of extracting the "cycle spectra" of a graph but in a different setting (basically, a histogram of the number of cycles of length `n`, with `n` being an integer), I can say that identifying ANY cycle within a dataset is an incredibly hard problem. 

To make it tractable, we would have to limit `n` within a reasonable range and possibly make the check "eventual" by pushing it out to its own thread. It is not something that you can do in a few milliseconds.

In any case, I have included this as an "enhancement" and see how we can incorporate this aspect into neomodel later on.

All the best
